axelrod/strategies/axelrod_second.py

```diff
@@ -1597,3 +1597,116 @@ def strategy(self, opponent: Player) -> Action:
         number_defects = np.sum(did_d(recent_history))
 
         return random_choice(self.prob_coop[number_defects])
+
+
+class RichardHufford(Player):
+    """
+    Strategy submitted to Axelrod's second tournament by Richard Hufford (K47R)
+    and came in sixteenth in that tournament.
+
+    The strategy tracks opponent "agreements," that is whenever the opponent's
+    previous move is the some as this player's move two turns ago.  If the
+    opponent's first move is a Defection, this is counted as a disagreement,
+    and otherwise an agreement.  From the agreement counts, two measures are
+    calculated:
+
+    - `proportion_agree`:  This is the number of agreements (through opponent's
+      last turn) + 2 divided by the current turn number.
+    - `last_four_num`:  The number of agreements in the last four turns.  If
+      there have been fewer than four previous turns, then this is number of
+      agreement + (4 - number of past turns).
+
+    We then use these measures to decide how to play, using these rules:
+
+    1. If `proportion_agree` > 0.9 and `last_four_num` >= 4, then Cooperate.
+    2. Otherwise if `proportion_agree` >= 0.625 and `last_four_num` >= 2, then
+       Tit-for-Tat.
+    3. Otherwise, Defect.
+
+    However, if the opponent has Cooperated the last `streak_needed` turns,
+    then the strategy deviates from the usual strategy, and instead Defects.
+    (We call such deviation an "aberration.")  In the turn immediately after an
+    aberration, the strategy doesn't override, even if there's a streak of
+    Cooperations.  Two turns after an aberration, the strategy:  Restarts the
+    Cooperation streak (never looking before this turn); Cooperates; and
+    changes `streak_needed` to:
+
+    floor(20.0 * `num_abb_def` / `num_abb_coop`) + 1
+
+    Here `num_abb_def` is 2 + the number of times that the opponent Defected in
+    the turn after an aberration, and `num_abb_coop` is 2 + the number of times
+    that the opponent Cooperated in response to an aberration.
+
+    Names:
+
+    - RichardHufford: [Axelrod1980b]_
+    """
+
+    name = 'RichardHufford'
+    classifier = {
+        'memory_depth': float("inf"),
+        'stochastic': False,
+        'makes_use_of': set(),
+        'long_run_time': False,
+        'inspects_source': False,
+        'manipulates_source': False,
+        'manipulates_state': False
+    }
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.num_agreements = 2
+        self.last_four_agreements = [1] * 4
+        self.last_four_index = 0
+
+        self.streak_needed = 21
+        self.current_streak = 2
+        self.last_aberration = float("inf")
+        self.num_abb_coop = 2
```

Comment:
This is short for "number of aberrations following cooperation" right? In which case I think the `abb` is confusing, perhaps:

```
ab_after_coop_count
ab_after_def_count
```

?

Happy for other suggestions.

Reply:
It's cooperations after aberration, so maybe coop_after_abb_count

Reply:
Cool, sounds good but a single `b` (not `abb`?): `coop_after_ab_count` and `def_after_ab_count`?

Reply:
Because I can't spell.  A single `b` is good.